crypto/src/main/java/org/springframework/security/crypto/password/DelegatingPasswordEncoder.java

```diff
@@ -245,12 +245,12 @@ private String extractId(String prefixEncodedPassword) {
 			return null;
 		}
 		int start = prefixEncodedPassword.indexOf(this.idPrefix);
-		if (start != 0) {
+		if (start == -1 && !prefixEncodedPassword.contains(this.idSuffix)) {
```

Comment:
I'd like to be a bit more conservative with changing this behavior, given that it is called by more than one flow in this class. What if the `matches` implementation itself did the needed checking? Something like this:

```java
String id = extractId(prefixEncodedPassword);
if (has a value) {
    throw no mapping exception
}
start = indexof idPrefix
end = indexof idSuffix, start
if (start < 0 && end < 0) {
    throw no encoder prefix 
}
throw malformed encoder prefix
```

In exchange for a small amount of duplication, we don't change the expectations of other flows calling `extractId`.